### PR TITLE
Seek FLV file and play RTMP

### PR DESCRIPTION
**Pull request by mekya**
FLV files can be seeked if keyframes exists as metatag.

RTMP streams can be played with RTMPDataSource

### CODE REVIEW

**Comment by mekya** [comment]
I did author these commits.

**Comment by ckarthikv** [comment]
@mekya - Just tried playing a live rtmp stream using the your commits using RTMP datasource - I got the following error below -

01-16 19:31:02.590 29424-30286/com.google.android.exoplayer.demo E/ExoPlayerImplInternal: Internal track renderer error.
                                                                                          com.google.android.exoplayer.ExoPlaybackException: com.google.android.exoplayer.ParserException
                                                                                              at com.google.android.exoplayer.SampleSourceTrackRenderer.maybeThrowError(SampleSourceTrackRenderer.java:154)
                                                                                              at com.google.android.exoplayer.SampleSourceTrackRenderer.maybeThrowError(SampleSourceTrackRenderer.java:145)
                                                                                              at com.google.android.exoplayer.ExoPlayerImplInternal.incrementalPrepareInternal(ExoPlayerImplInternal.java:283)
                                                                                              at com.google.android.exoplayer.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:205)
                                                                                              at android.os.Handler.dispatchMessage(Handler.java:98)
                                                                                              at android.os.Looper.loop(Looper.java:135)
                                                                                              at android.os.HandlerThread.run(HandlerThread.java:61)
                                                                                              at com.google.android.exoplayer.util.PriorityHandlerThread.run(PriorityHandlerThread.java:40)
                                                                                           Caused by: com.google.android.exoplayer.ParserException
                                                                                              at com.google.android.exoplayer.extractor.flv.ScriptTagPayloadReader.parsePayload(ScriptTagPayloadReader.java:85)
                                                                                              at com.google.android.exoplayer.extractor.flv.TagPayloadReader.consume(TagPayloadReader.java:87)
                                                                                              at com.google.android.exoplayer.extractor.flv.FlvExtractor.readTagData(FlvExtractor.java:251)
                                                                                              at com.google.android.exoplayer.extractor.flv.FlvExtractor.read(FlvExtractor.java:146)
                                                                                              at com.google.android.exoplayer.extractor.ExtractorSampleSource$ExtractingLoadable.load(ExtractorSampleSource.java:753)
                                                                                              at com.google.android.exoplayer.upstream.Loader$LoadTask.run(Loader.java:209)
                                                                                              at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
                                                                                              at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                                              at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
                                                                                              at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
                                                                                              at java.lang.Thread.run(Thread.java:818)

File: README.md [hunk 1]
```diff
@@ -1,5 +1,13 @@
 # ExoPlayer Readme #
 
+### Exoplayer is extended
```
Comment (ojw28) [inline README.md]: This should not be a part of this change.

File: demo/src/main/java/com/google/android/exoplayer/demo/player/ExtractorRendererBuilder.java [hunk 2]
```diff
@@ -59,8 +60,9 @@ public void buildRenderers(DemoPlayer player) {
     DefaultBandwidthMeter bandwidthMeter = new DefaultBandwidthMeter(player.getMainHandler(),
         null);
     DataSource dataSource = new DefaultUriDataSource(context, bandwidthMeter, userAgent);
+    //DataSource dataSource = new RtmpDataSource();
     ExtractorSampleSource sampleSource = new ExtractorSampleSource(uri, dataSource, allocator,
-        BUFFER_SEGMENT_COUNT * BUFFER_SEGMENT_SIZE);
+        BUFFER_SEGMENT_COUNT * BUFFER_SEGMENT_SIZE, new FlvExtractor());
```
Comment (ojw28) [inline demo/src/main/java/com/google/android/exoplayer/demo/player/ExtractorRendererBuilder.java]: Ditto

**Comment by ojw28** [comment]
This really needs to be split up into two distinct pull requests, since it's delivering two separate things. Please could you open 2x new requests, one for FLV and one for RTMP? For the FLV one, can you please provide a link to some sample media that we can use to test the implementation in the pull request description? Thanks.

File: demo/src/main/java/rtmp_client/RtmpClient.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package net.butterflytv.rtmp_client;
```
Comment (ojw28) [inline demo/src/main/java/rtmp_client/RtmpClient.java]: We can't accept a change that includes a random package like this. Where does this file actually originate from? If you copied it from some other project, note that you can't submit it here claiming to be the author unless you were also the original author of the original file. Please clarify. Thanks!

**Comment by Android4MediaPlayer** [comment]
@mekya where is the code of librtmp_jni.so?

**Comment by mekya** [comment]
I am gonna share librtmp_jni as a project on github. It is based on librtmp. Moreover, I gonna make other things clear 

Sorry for the delay. Working hard lately. 

@ojw28 @ckarthikv @Android4MediaPlayer

**Comment by ojw28** [comment]
Thanks! It's definitely worth splitting the FLV change into a separate request, since that one looks a lot more straightforward for us to merge than the RTMP change.

**Comment by mekya** [comment]
Thank you. Ok I am gonna handle it.

**Comment by ckarthikv** [comment]
@mekya - The sample works well for me, but somehow a custom live RTMP source doesn't seem to work with your changes. Any recommendations? I tried adding "?live=1" at the end of the RTMP url also, but only audio track is being played and not the video track (although the stream has both video and audio)

**Comment by mekya** [comment]
@ckarthikv I recommend you not to add question mark. Add something like that
"rtmp://www.butterflytv.net/streamname live=1"

this is librtmp specification.  We have tested live streaming lots of times through Red5 and probably puting the app production in these days. 

BTW, some rtmp servers does not indicate  that audio or video exist in FLV header.  ExoPlayer implements the specification strictly so if FLV header does not indicate video exists, ExoPlayer does not play the video. Most of the other media players does not care too much flv header. They play audio or video if they encounter frames while reading the stream. 

To sum up, I have come across the same issue.  Make sure that FLV header is correct. 

Forked repo is here https://github.com/ButterflyTV/ExoPlayer-with-RTMP-and-FLV-seek

**Comment by developer-int** [comment]
@mekya  I did encountered same issue , while i try playing livestream from Red5
it gets disconnected automatically
i am using below player :  https://github.com/ButterflyTV/ExoPlayer-with-RTMP-and-FLV-seek

Below are my Red5 Logs :

[INFO] [Red5_Scheduler_Worker-21] app.Application - W3C x-category:stream x-
event:play c-ip:192.168.0.4 x-sname:e372cde8-7d0c-4135-8f75-97e939b91f79 x-name:
1234_801cf519aca6ebd42a13aca73f6b1ef8
[INFO] [Red5_Scheduler_Worker-17] app.Application - W3C x-category:stream x-
event:play c-ip:192.168.0.4 x-sname:e372cde8-7d0c-4135-8f75-97e939b91f79
[INFO] [NioProcessor-12] app.Application - W3C x-category:session x-event:di
sconnect c-ip:192.168.0.4 c-client-id:90
[INFO] [Red5_Scheduler_Worker-6] app.Application - W3C x-category:stream x-e
vent:stop c-ip:192.168.0.4 cs-bytes:3510 sc-bytes:630499 x-sname:e372cde8-7d0c-4
135-8f75-97e939b91f79

**Comment by mekya** [comment]
@developer-int Player needs to be debugged with your stream. 

Put breakpoints to some points in [FLVExtractor](https://github.com/ButterflyTV/ExoPlayer-with-RTMP-and-FLV-seek/blob/master/library/src/main/java/com/google/android/exoplayer/extractor/flv/FlvExtractor.java) and try to understand if problem is there.

**Comment by developer-int** [comment]
@mekya 
its working now , i was using customized url to publish/play my content
rtmp://[ip]/[appame]/[custompath]/[streamname]  live=1
but it wasn't working
Changed to 
rtmp://[ip]/[appame]/[streamname]  live=1
Its working now
Strange problem, i had to change my publish and play url, but fixed now
Thanks for getting back to me

**Comment by mekya** [comment]
great to hear that :)

You're welcome ;)

@developer-int

**Comment by developer-int** [comment]
@mekya 
Hi , i have encountered another issue , FLV file is not playing
here is link to file , i am trying to play
http://androidmobiapps.com/12345.flv

**Comment by ardoramor** [comment]
Are there plans to merge this PR?

**Comment by mekya** [comment]
if I have some time, I can change somethings and make a new PR.

**Comment by ardoramor** [comment]
@mekya Is there any way in which I can help? I can find the time after next two weeks to contribute.

**Comment by mekya** [comment]
Thank you @ardoramor 
There are two things
- You can make a new PR that canbe accepted
- I don't know if there is a plan to support webrtc but this is something I really want to add. Maybe we can do something together. 

How does it sound?
